pkg/pillar/controllerconn/resolver.go

```diff
@@ -380,25 +411,62 @@ func ResolveWithPortsLambda(domain string,
 		close(quit)
 		<-wgChan
 	}()
-	select {
-	case <-wgChan:
-		var responses []DNSResponse
-		if countDNSRequests == 0 {
-			// fallback in case no resolver is configured
-			ips, err := net.LookupIP(domain)
-			if err != nil {
-				return nil, append(errs, fmt.Errorf("fallback resolver failed: %+v", err))
+
+	var responses []DNSResponse
+	processResponses := func(dnsResponses []DNSResponse) {
+		for _, dnsResp := range dnsResponses {
+			if dnsResp.IP.To4() != nil {
+				waitForIPv4 = false
+			} else {
+				waitForIPv6 = false
+			}
+			// Avoid duplicates.
+			var duplicate bool
+			for i := range responses {
+				if responses[i].IP.Equal(dnsResp.IP) {
+					duplicate = true
+					break
+				}
+			}
+			if !duplicate {
+				responses = append(responses, dnsResp)
```

Comment:
Do we want to record duplicates or it's not important information for us?

Reply:
We do not need duplicate entries.
This here is used to resolve and cache controller IPv4/IPv6 address -- no need to cache the same IP address multiple times.

Reply:
Makes sense, thanks!